File: go/federation_bindings_test.go
package realmid

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestFederationBindings_CRUD(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/auth/login", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{"status": "ok", "subject_type": "platform", "refresh_token": "rtok-platform", "access_token": "ptok", "expires_in": 300})
	})
	base := "/platforms/" + testRealmID + "/federation-bindings"
	var createBody map[string]any
	var listMethod, createMethod, delMethod, delPath string
	mux.HandleFunc(base, func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case "GET":
			listMethod = r.Method
			_ = json.NewEncoder(w).Encode(map[string]any{
				"items": []any{
					map[string]any{"id": "fb1", "issuer": "https://token.actions.githubusercontent.com", "status": "active", "match_claims": map[string]any{"repository": "acme/billing"}},
				},
				"next_cursor": "",
			})
		case "POST":
			createMethod = r.Method
			_ = json.NewDecoder(r.Body).Decode(&createBody)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"id": "fb2", "platform_id": testRealmID, "issuer": createBody["issuer"],
				"audience": "ri-const", "status": "active", "mapped_role": "platform_api",
				"match_claims": createBody["match_claims"], "scope": []any{"read"},
			})
		}
	})
	mux.HandleFunc(base+"/fb2", func(w http.ResponseWriter, r *http.Request) {
		delMethod, delPath = r.Method, r.URL.Path
		_ = json.NewEncoder(w).Encode(map[string]any{"status": "revoked", "id": "fb2"})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})

	// List
	page, err := r.FederationBindings.List(context.Background()).Page(context.Background(), nil)
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if listMethod != "GET" || len(page.Items) != 1 || page.Items[0].ID != "fb1" || page.Items[0].MatchClaims["repository"] != "acme/billing" {
		t.Errorf("unexpected list: %s %+v", listMethod, page.Items)
	}

	// Create
	fb, err := r.FederationBindings.Create(context.Background(), FederationBindingCreate{
		Issuer:      "https://token.actions.githubusercontent.com",
		MatchClaims: map[string]string{"repository": "acme/billing"},
		MappedRole:  "platform_api",
		Scope:       []string{"read"},
	})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if createMethod != "POST" || fb.ID != "fb2" || fb.Audience != "ri-const" || fb.Status != "active" {
		t.Errorf("unexpected create: %s %+v", createMethod, fb)
	}
	if createBody["issuer"] != "https://token.actions.githubusercontent.com" {
		t.Errorf("create body issuer: %v", createBody["issuer"])
	}
	mc, _ := createBody["match_claims"].(map[string]any)
	if mc == nil || mc["repository"] != "acme/billing" {
		t.Errorf("create body match_claims: %v", createBody["match_claims"])
	}

	// Revoke
	res, err := r.FederationBindings.Revoke(context.Background(), "fb2")
	if err != nil {
		t.Fatalf("Revoke: %v", err)
	}
	if delMethod != "DELETE" || delPath != base+"/fb2" || res.Status != "revoked" || res.ID != "fb2" {
		t.Errorf("unexpected revoke: %s %s %+v", delMethod, delPath, res)
	}
}

File: ts/src/federation-bindings.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { createRealm } from "./realm.js";

function loginOr(handler: (url: string, init?: RequestInit) => Response): typeof fetch {
  return (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    if (url.endsWith("/auth/login")) {
      return new Response(JSON.stringify({ status: "ok", subject_type: "platform", refresh_token: "rtok-platform", access_token: "pt_x", expires_in: 300 }), {
        status: 200, headers: { "content-type": "application/json" },
      });
    }
    return handler(url, init);
  }) as typeof fetch;
}

test("federationBindings.list: pages the platform bindings", async () => {
  let hitUrl = "";
  const fetch = loginOr((url) => {
    hitUrl = url;
    return new Response(JSON.stringify({
      items: [{ id: "fb1", issuer: "https://token.actions.githubusercontent.com", status: "active", match_claims: { repository: "acme/billing" } }],
      next_cursor: null,
    }), { status: 200, headers: { "content-type": "application/json" } });
  });
  const realm = createRealm({ realmId: "r-1", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const p = await realm.federationBindings.list().page();
  assert.match(hitUrl, /\/platforms\/r-1\/federation-bindings/);
  assert.equal(p.items[0]!.id, "fb1");
  assert.equal(p.items[0]!.match_claims.repository, "acme/billing");
});

test("federationBindings.create: POSTs snake_case body and decodes", async () => {
  let hitMethod = "";
  let hitBody: Record<string, unknown> = {};
  const fetch = loginOr((_url, init) => {
    if (init?.method) hitMethod = init.method;
    if (init?.body) hitBody = JSON.parse(init.body as string);
    return new Response(JSON.stringify({ id: "fb2", platform_id: "r-1", issuer: "https://token.actions.githubusercontent.com", audience: "ri-const", status: "active", mapped_role: "platform_api" }),
      { status: 201, headers: { "content-type": "application/json" } });
  });
  const realm = createRealm({ realmId: "r-1", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const fb = await realm.federationBindings.create({
    issuer: "https://token.actions.githubusercontent.com",
    matchClaims: { repository: "acme/billing" },
    mappedRole: "platform_api",
    scope: ["read"],
  });
  assert.equal(hitMethod, "POST");
  assert.equal(fb.id, "fb2");
  assert.equal(fb.audience, "ri-const");
  assert.deepEqual(hitBody.match_claims, { repository: "acme/billing" });
  assert.equal(hitBody.mapped_role, "platform_api");
  assert.deepEqual(hitBody.scope, ["read"]);
});

test("federationBindings.revoke: DELETEs the binding by id", async () => {
  let hitUrl = "";
  let hitMethod = "";
  const fetch = loginOr((url, init) => {
    hitUrl = url;
    if (init?.method) hitMethod = init.method;
    return new Response(JSON.stringify({ status: "revoked", id: "fb2" }),
      { status: 200, headers: { "content-type": "application/json" } });
  });
  const realm = createRealm({ realmId: "r-1", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const res = await realm.federationBindings.revoke("fb2");
  assert.equal(hitMethod, "DELETE");
  assert.match(hitUrl, /\/platforms\/r-1\/federation-bindings\/fb2$/);
  assert.equal(res.status, "revoked");
  assert.equal(res.id, "fb2");
});
